packages/backend/src/api/controllers/ForcedTradeOfferController.ts

```diff
@@ -0,0 +1,289 @@
+import { encodeAssetId } from '@explorer/encoding'
+import { AssetId, EthereumAddress, StarkKey, Timestamp } from '@explorer/types'
+import {
+  recoverAddress,
+  solidityKeccak256,
+  solidityPack,
+} from 'ethers/lib/utils'
+
+import {
+  ForcedTradeAcceptRecord,
+  ForcedTradeInitialOfferRecord,
+  ForcedTradeOfferRepository,
+} from '../../peripherals/database/ForcedTradeOfferRepository'
+import {
+  PositionRecord,
+  StateUpdateRepository,
+} from '../../peripherals/database/StateUpdateRepository'
+import { UserRegistrationEventRepository } from '../../peripherals/database/UserRegistrationEventRepository'
+import { ControllerResult } from './ControllerResult'
+
+export class ForcedTradeOfferController {
+  constructor(
+    private offerRepository: ForcedTradeOfferRepository,
+    private stateUpdateRepository: StateUpdateRepository,
+    private userRegistrationEventRepository: UserRegistrationEventRepository
+  ) {}
+
+  async postOffer(
+    offer: Omit<ForcedTradeInitialOfferRecord, 'createdAt' | 'id'>
+  ): Promise<ControllerResult> {
+    const positionA = await this.stateUpdateRepository.getPositionById(
+      offer.positionIdA
+    )
+
+    if (!positionA) {
+      return { type: 'not found', content: 'Position does not exist.' }
+    }
+
+    const frozenBalance = await this.calculateFrozenBalance(
+      offer.starkKeyA,
+      offer.aIsBuyingSynthetic ? AssetId.USDC : offer.syntheticAssetId
+    )
+
+    const offerValidated = validateInitialOffer(offer, positionA, frozenBalance)
+
+    if (!offerValidated) {
+      return { type: 'bad request', content: 'Your offer is invalid.' }
+    }
+
+    const createdAt = Timestamp(Date.now())
+
+    const id = await this.offerRepository.addInitialOffer({
+      createdAt,
+      ...offer,
+    })
+    return { type: 'created', content: { id } }
+  }
+
+  async acceptOffer(
+    initialOfferId: number,
+    acceptOfferData: Omit<ForcedTradeAcceptRecord, 'acceptedAt'>
+  ): Promise<ControllerResult> {
+    const positionB = await this.stateUpdateRepository.getPositionById(
+      acceptOfferData.positionIdB
+    )
+    const userRegistrationEventB =
+      await this.userRegistrationEventRepository.findByStarkKey(
+        acceptOfferData.starkKeyB
+      )
+
+    if (!positionB || !userRegistrationEventB) {
+      return { type: 'not found', content: 'Position does not exist.' }
+    }
+    const initialOffer = await this.offerRepository.findInitialOfferById(
+      initialOfferId
+    )
+
+    if (!initialOffer) {
+      return { type: 'not found', content: 'Offer does not exist.' }
+    }
+
+    const accceptOffer = await this.offerRepository.findAcceptOfferById(
+      initialOfferId
+    )
+
+    if (accceptOffer) {
+      return {
+        type: 'bad request',
+        content: 'Offer already accepted by a user.',
+      }
+    }
+
+    const frozenBalance = await this.calculateFrozenBalance(
```

Comment:
When someone agrees to trade they don't actually need to have any of the assets that the other party wants.